security/pkg/nodeagent/sds/sdsservice_test.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package sds
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"testing"
+	"time"
+
+	api "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	authapi "github.com/envoyproxy/go-control-plane/envoy/api/v2/auth"
+	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	sds "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/gogo/protobuf/types"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"k8s.io/apimachinery/pkg/util/uuid"
+)
+
+var (
+	socket               string
+	fakeCertificateChain = []byte{01}
+	fakePrivateKey       = []byte{02}
+)
+
+func TestStreamSecrets(t *testing.T) {
+	socket = fmt.Sprintf("/tmp/gotest%q.sock", string(uuid.NewUUID()))
+	testHelper(t, socket, sdsRequestStream)
+}
+
+func TestFetchSecrets(t *testing.T) {
+	socket = fmt.Sprintf("/tmp/gotest%q.sock", string(uuid.NewUUID()))
+	testHelper(t, socket, sdsRequestFetch)
+}
+
+type secretCallback func(string, *api.DiscoveryRequest) (*api.DiscoveryResponse, error)
+
+func testHelper(t *testing.T, testSocket string, cb secretCallback) {
+	arg := Args{
+		SDSUdsSocket: testSocket,
+	}
+	st := &mockSecretStore{}
+	server, err := NewServer(arg, st)
+	defer server.Close()
+
+	if err != nil {
+		t.Fatalf("failed to start grpc server for sds: %v", err)
+	}
+
+	req := &api.DiscoveryRequest{
+		ResourceNames: []string{"test"},
+		Node: &core.Node{
+			Id: "sidecar~127.0.0.1~id~local",
+		},
+	}
+
+	wait := 300 * time.Millisecond
+	for try := 0; try < 5; try++ {
+		time.Sleep(wait)
```

Comment:
Move the sleep to the end of the loop? So that we don't have to wait for the first iteration?

Reply:
as mentioned in another comment, sleep here give server some buffer to get ready to serve traffic.